### PR TITLE
Make metadata properties more efficient

### PR DESCRIPTION
**Pull request by panopticoncentral**
In heap dumps of large CPS solutions, ~15% of the heap is taken up by VisualStudioProject instances. The issue is VisualStudioProject._allMetadataReferences. While most metadata references have the default metadata properties, every time a metadata reference is added the code currently adds an entry in _allMetadataReferences pointing to a list of metadata properties. First, if the metadata properties are default and there's only one reference, we're wasting space by allocating the list in the first place. But second, this list starts with an initial capacity of 4, which means that we're wasting quadruple the space. For every metadata reference. Of which there are a _lot_ in .NET Core.

The solution is just to optimize the case where there is only one of the metadata reference and it has the default properties. So:

* `!_allMetadataReferences.TryGetValue(path, out var list)` means "we have not seen this reference"
* `_allMetadataReferences.TryGetValue(path, out var list) && list == null` means "we have seen only one of this reference and its metadata properties are the default ones."
* Otherwise, we've either seen this reference multiple times, or the one time we saw it its metadata properties were not the default.

I chose not to optimize the remove path (i.e. you have multiple references and you remove every one except the default properties) because it seemed to be a pretty esoteric situation.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
Is it worth adding a comment to _allMetadataReferences to mention the special logic here? Also, how big were those dictionaries even after this? If we don't have a batch going, we might be able to drop the dictionaries entirely and just fall back to using what is inside of the current workspace snapshot.

**Comment by panopticoncentral** [comment]
@jasonmalinowski The VisualStudioProject dropped off the heap dump top list once I did this. When opening a large solution (like, 520 projects) I stopped counting around 65k metadata references. So there are a lot. Let me know if you think we should do something else, otherwise I can add comments to clarify the situation better...

**Comment by jasonmalinowski** [comment]
My only alternative thought was to replace the dictionary from being a Dictionary<string, List<MetadataReferenceProperties>> to a Dictionary<string, ImmutableArray<MetadataReferenceProperties>>, and then you can just share the singleton instance of the ImmutableArray everywhere. It means the add/removal logic is still more complicated, but since we're having to make new extension methods for this -- maybe that's still OK?

**Comment by jasonmalinowski** [comment]
So looking I'd either say do the ImmutableArray-thats-shared approach, or just add more comments. I worry about the next person who will have to figure out the MultiAddWithDefault and wonder what it's doing...

**Comment by CyrusNajmabadi** [comment]
I agree wth jason.  I think the 'WithDefault' is trying to hard to save that last 0.1% of memory. I think from sharing the immutable array you'll get the lions share (pun intended), while keeping things much clearer.

**Comment by panopticoncentral** [comment]
Does this look better @CyrusNajmabadi and @jasonmalinowski ?

File: src/Workspaces/Core/Portable/Utilities/IDictionaryExtensions.cs [hunk 1]
```diff
@@ -42,6 +43,16 @@ internal static class IDictionaryExtensions
...
         }
 
+        public static void MultiAdd<TKey, TValue>(this IDictionary<TKey, ImmutableArray<TValue>> dictionary, TKey key, TValue value)
+        {
+            if (!dictionary.TryGetValue(key, out var collection))
+            {
+                collection = ImmutableArray<TValue>.Empty;
+            }
+
+            dictionary[key] = collection.Add(value);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Utilities/IDictionaryExtensions.cs]: Did we lose the special case for the single-element sharing though?

Comment (panopticoncentral) [reply src/Workspaces/Core/Portable/Utilities/IDictionaryExtensions.cs]: Oh, yeah, right. Updated.

**Comment by jasonmalinowski** [comment]
Tagging @jinujoseph then for the M2 approval for a good perf fix.

File: src/Workspaces/Core/Portable/Utilities/IDictionaryExtensions.cs [hunk 2]
```diff
@@ -42,6 +43,16 @@ internal static class IDictionaryExtensions
...
         }
 
+        public static void MultiAdd<TKey, TValue>(this IDictionary<TKey, ImmutableArray<TValue>> dictionary, TKey key, TValue value, ImmutableArray<TValue> defaultArray)
+        {
+            if (!dictionary.TryGetValue(key, out var collection))
+            {
+                collection = ImmutableArray<TValue>.Empty;
+            }
+
+            dictionary[key] = collection.IsEmpty && value == default ? defaultArray : collection.Add(value);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Utilities/IDictionaryExtensions.cs]:
```suggestion
            dictionary[key] = collection.IsEmpty && value == defaultArray[0] ? defaultArray : collection.Add(value);
```

@panopticoncentral One small last suggestion (sorry) but hopefully you can just hit apply: this is mostly to guard against if somebody handed in an array that *wasn't* filled with a default element, we'd still be functionally correct. It also means that if we ever have that case down the road it'll work too if the non-default value is common.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs [hunk 3]
```diff
@@ -17,6 +17,8 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
 {
     internal sealed class VisualStudioProject
     {
+        private static readonly ImmutableArray<MetadataReferenceProperties> s_defaultMetadataReferenceProperties = ImmutableArray<MetadataReferenceProperties>.Empty.Add(default);
```
Comment (Therzok) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs]:
```suggestion
        private static readonly ImmutableArray<MetadataReferenceProperties> s_defaultMetadataReferenceProperties = ImmutableArray.Create<MetadataReferenceProperties>(default);
```

**Pull request merged**